compiler/rustc_middle/src/mir/interpret/pointer.rs

```diff
@@ -141,6 +126,80 @@ pub trait Provenance: Copy + fmt::Debug + 'static {
     fn join(left: Option<Self>, right: Option<Self>) -> Option<Self>;
 }
 
+/// The type of provenance in the compiler-time interpreter.
+/// This is a packed representation of an `AllocId` and an `immutable: bool`.
+#[derive(Copy, Clone, Eq, Hash, Ord, PartialEq, PartialOrd)]
+pub struct CtfeProvenance(NonZeroU64);
+
+impl From<AllocId> for CtfeProvenance {
+    fn from(value: AllocId) -> Self {
+        let prov = CtfeProvenance(value.0);
+        assert!(!prov.immutable(), "`AllocId` with the highest bit set cannot be used in CTFE");
```

Comment:
Everything else in this PR is mostly just mechanical changes or pretty simple, but I do not understand what we're doing here. Why can we only create a `CtfeProvenance` that we can write through?

Reply:
This is asserting that `value` is less than 2^63, to ensure we can use the highest bit to store "immutable".

Reply:
I see. I would have written this check without going through the accessor, but this is fine. The assert message should have clued me in.